mpfilepicker/src/iosMain/kotlin/com/darkrockstudios/libraries/mpfilepicker/IosFilePickerLauncher.kt

```diff
@@ -0,0 +1,172 @@
+import kotlin.native.concurrent.ThreadLocal
+
+/**
+ * Wraps platform specific implementation for launching a
+ * File Picker.
+ *
+ * @param initialDirectory Initial directory that the
+ *  file picker should open to.
+ * @param pickerMode [Mode] to open the picker with.
+ *
+ */
+public class FilePickerLauncher(
+	private val initialDirectory: String?,
+	private val pickerMode: Mode,
+	private val onFileSelected: FileSelected,
+) {
+
+	@ThreadLocal
+	public companion object {
+		/**
+		 * For use only with launching plain (no compose dependencies)
+		 * file picker. When a function completes iOS deallocates
+		 * unreferenced objects created within it, so we need to
+		 * keep a reference of the active launcher.
+		 */
+		internal var activeLauncher: FilePickerLauncher? = null
+	}
+
+	/**
+	 * Identifies the kind of file picker to open. Either
+	 * [Directory] or [File].
+	 */
+	public sealed interface Mode {
+		/**
+		 * Use this mode to open a [FilePickerLauncher] for selecting
+		 * folders/directories.
+		 */
+		public data object Directory : Mode
+
+		/**
+		 * Use this mode to open a [FilePickerLauncher] for selecting
+		 * files.
+		 *
+		 * @param extensions List of file extensions that can be
+		 *  selected on this file picker.
+		 */
+		public data class File(val extensions: List<String>) : Mode
+	}
+
+	private val pickerDelegate = object : NSObject(),
+		UIDocumentPickerDelegateProtocol,
+		UIAdaptivePresentationControllerDelegateProtocol {
+
+		override fun documentPicker(
+			controller: UIDocumentPickerViewController, didPickDocumentsAtURLs: List<*>
+		) {
+			(didPickDocumentsAtURLs.firstOrNull() as? NSURL).let { selected ->
+				onFileSelected(selected?.path?.let { IosFile(it, selected) })
+			}
+		}
+
+		override fun documentPickerWasCancelled(
+			controller: UIDocumentPickerViewController
+		) {
+			onFileSelected(null)
+		}
+
+		override fun presentationControllerWillDismiss(
+			presentationController: UIPresentationController
+		) {
+			(presentationController.presentedViewController as? UIDocumentPickerViewController)
+				?.let { documentPickerWasCancelled(it) }
+		}
+	}
+
+	private val contentTypes: List<UTType>
+		get() = when (pickerMode) {
+			is Directory -> listOf(UTTypeFolder)
+			is File -> pickerMode.extensions
+				.mapNotNull { UTType.typeWithFilenameExtension(it) }
+				.ifEmpty { listOf(UTTypeContent) }
+		}
+
+	private fun createPicker() = UIDocumentPickerViewController(
+		forOpeningContentTypes = contentTypes
+	).apply {
+		delegate = pickerDelegate
+		initialDirectory?.let { directoryURL = NSURL(string = it) }
+	}
+
+	public fun launchFilePicker() {
+		activeLauncher = this
+
+		UIApplication.sharedApplication.keyWindow?.rootViewController?.presentViewController(
+			// Reusing a closed/dismissed picker causes problems with
+			// triggering delegate functions, launch with a new one.
+			createPicker(),
+			animated = true,
+			completion = null
+		)
+	}
+}
+
+public suspend fun launchFilePicker(
```

Comment:
Single function file picker. Coroutines might just be necessary if we want to return the result instead of a callback?

Reply:
I did not get the question, but if you are asking about keeping the function as suspend or not, then I think it is okay. In JS target those functions are also suspend.